src/Infrastructure/Rsp.IrasService.Infrastructure/Interceptors/AuditTrailInterceptor.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+using Rsp.IrasService.Domain.Entities;
+using Rsp.IrasService.Domain.Interfaces;
+using Rsp.IrasService.Infrastructure.Helpers;
+
+namespace Rsp.IrasService.Infrastructure.Interceptors;
+
+[ExcludeFromCodeCoverage]
+public class AuditTrailInterceptor(IAuditTrailDetailsService auditTrailDetailsService, IEnumerable<IAuditTrailHandler> auditTrailHandlers) : SaveChangesInterceptor
+{
+    public override async ValueTask<InterceptionResult<int>> SavingChangesAsync(
+        DbContextEventData eventData,
+        InterceptionResult<int> result,
+        CancellationToken cancellationToken = default)
+    {
+        if (eventData.Context is not IrasContext dbContext)
+        {
+            return await base.SavingChangesAsync(eventData, result, cancellationToken);
+        }
+
+        var auditableEntries = dbContext
+            .ChangeTracker
+            .Entries<IAuditable>();
+
+        if (!auditableEntries.Any())
+        {
+            return await base.SavingChangesAsync(eventData, result, cancellationToken);
+        }
+
+        var systemAdminEmail = auditTrailDetailsService.GetEmailFromHttpContext();
+
+        var auditTrailRecords = new List<ReviewBodyAuditTrail>();
+
+        foreach (var entry in auditableEntries)
+        {
+            foreach (var auditTrailHandler in auditTrailHandlers)
+            {
+                if (auditTrailHandler.CanHandle(entry.Entity))
+                {
+                    auditTrailRecords.AddRange(auditTrailHandler.GenerateAuditTrails(entry, systemAdminEmail));
+                }
+            }
```

Comment:
it's better to use LINQ and filter out the handlers upfront.
```suggestion
            var handlers = auditTrailHandlers
                                .Where(auditTrailHandler => auditTrailHandler.CanHandle(entry.Entity))
                                .ToList();

            foreach (var auditTrailHandler in handlers)
            {
                auditTrailRecords.AddRange(auditTrailHandler.GenerateAuditTrails(entry, systemAdminEmail));
            }
```